src/rosinstall_generator/distro.py

```diff
@@ -132,26 +133,38 @@ def _generate_rosinstall_for_package(distro, pkg_name, flat=False, tar=False):
 
 
 def _generate_rosinstall(local_name, url, release_tag, tar=False, vcs_type=None):
+    logger = logging.getLogger('rosinstall_generator.generate')
+
     if tar:
-        # the repository name might be different than repo.name coming from rosdistro
-        repo_name = os.path.basename(url[:-4])
-        suffix = '.git'
-        if not url.endswith(suffix):
-            raise RuntimeError("The repository '%s' must end with '%s': %s" % (repo_name, suffix, url))
-        url = url[:-len(suffix)] + '/archive/{0}.tar.gz'.format(release_tag)
-        data = [{
-            'tar': {
-                'local-name': local_name,
-                'uri': url,
-                'version': '{0}-{1}'.format(repo_name, release_tag.replace('/', '-'))
-            }
-        }]
-    else:
-        data = [{
-            vcs_type or 'git': {
-                'local-name': local_name,
-                'uri': url,
-                'version': release_tag
+        # Github tarball:    https://github.com/ros/ros_comm/archive/1.11.20.tar.gz
+        # Bitbucket tarball: https://bitbucket.org/osrf/gazebo/get/gazebo7_7.3.1.tar.gz
+        # Gitlab tarball:    https://gitlab.com/gitlab-org/gitlab-ce/repository/archive.tar.gz?ref=master
+        match = re.match('(?:https?:\/\/|ssh:\/\/|git:\/\/|git@)([\w.-]+)[:/]([\w/-]*)(?:\.git)?$', url)
```

Comment:
I agree that it's fragile. Unfortunately, urlparse doesn't do the right thing with SSH urls:

```
>>> urlparse("https://github.com/example/riverdale-release.git")
ParseResult(scheme='https', netloc='github.com', path='/example/riverdale-release.git', params='', query='', fragment='')
>>> urlparse("git@gitlab.example.com:example/riverdale-release.git")
ParseResult(scheme='', netloc='', path='git@gitlab.example.com:example/riverdale-release.git', params='', query='', fragment='')
>>> urlparse("ssh://git@gitlab.example.com:example/riverdale-release.git")
ParseResult(scheme='ssh', netloc='git@gitlab.example.com:example', path='/riverdale-release.git', params='', query='', fragment='')
```

We could conceivably work with this, but it would be some effort to pre- and post-process for it.